tensorflow/python/lib/io/file_io.py

```diff
@@ -336,24 +339,29 @@ def write_string_to_file(filename, file_content):
 
 
 @tf_export(v1=["gfile.Glob"])
-def get_matching_files(filename):
+def get_matching_files(filename, num_threads=1):
   """Returns a list of files that match the given pattern(s).
 
   Args:
     filename: string or iterable of strings. The glob pattern(s).
+    num_threads: int. If `filename` is iterable of strings and
+      `num_threads` is greater than one, it will create a threadpool to
+      match all strings parallelly. The size of threadpool is
+      `min(num_threads, len(list(filename)))`.
 
   Returns:
     A list of strings containing filenames that match the given pattern(s).
 
   Raises:
   *  errors.OpError: If there are filesystem / directory listing errors.
   *  errors.NotFoundError: If pattern to be matched is an invalid directory.
+  *  TypeError: If any argument does not have the expected type.
   """
-  return get_matching_files_v2(filename)
+  return get_matching_files_v2(filename, num_threads)
 
 
 @tf_export("io.gfile.glob")
-def get_matching_files_v2(pattern):
+def get_matching_files_v2(pattern, num_threads=1):
```

Comment:
I thought we were going to automatically determine the best number of threads if user didn't supply any.

Reply:
Have done.

Reply:
Sorry that I reopened this conversation. As for the default values of `num_threads`, I think we may need to discuss it again.

I have made [benchmarks](https://github.com/tensorflow/tensorflow/pull/44269#issuecomment-721501440) on a multi-core (80 threads) server and the default value of `num_threads` (`len(patterns)`) can produce a quite positive optimization effect. However, when I apply this to another machine with poorer performance, it may produces negative effect and even sometimes makes the process hang. I think the cause is excessive CPU resource preemption with the parallelism enhanced at the C++ level (#44310).

**Therefore, considering of better versatility, I prefer to set the default value of `num_threads` to `1`**, i.e. turn [here](https://github.com/tensorflow/tensorflow/blob/d8192d06e91ce3be15e5ff8b95641bfe76226803/tensorflow/python/lib/io/file_io.py#L435) to `num_threads = num_threads or 1`, which can keep consistent with the previous codes by default, and work positively if user set it manually. In addition, we can explain the risk and recommended value of this parameter in the comments.

Look forward to your suggestions.